src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+  void checkTemp() {
+    var temperature = _temperature;
+    if (temperature != null) {
+      print('Ready to serve ' + temperature + '!');
+    }
+  }
+```
+
+Since the type promotion does apply to locals, this now works fine. If you need
+to *change* the value, just remember to store back to the field and not just the
+local.
+
+### Nullability and generics
+
+Like most modern statically-typed languages, Dart has generic classes and
+generic methods. They interact with nullability in a few ways that seem
+counter-intuitive but make sense once you think through the implications. First
+is that "is this type nullable?" is no longer a simple yes or no question.
+Consider:
+
+```dart
+class Box<T> {
+  final T object;
+  Box(this.object);
+}
+
+main() {
+  Box<String>("a string");
+  Box<int?>(null);
+}
+```
+
+In the definition of `Box`, is `T` a nullable type or a non-nullable type? As
+you can see, it can be instantiated with either kind. The answer is that `T` a
+*potentially nullable type*. Inside the body of a generic class or method, a
+potentially nullable type has all of the restrictions of both nullable types
+*and* non-nullable types.
+
+The former means you can't call any methods on it except the handful defined on
+Object. The latter means that you must initialize any fields or variables of
+that type before they're used. This can make type parameters pretty hard to work with.
+
+In practice, a few patterns show up. In collection-like classes where the type
+parameter can be instantiated with any type at all, you just have to deal with
+the restrictions. In most cases, like the example here, it means ensuring you do
+have access to a value of the type argument's type whenever you need to work
+with one. Fortunately, collection-like classes rarely call methods on their
+elements.
+
+In places where you don't have access to a value, you can make the type
+parameter nullable:
+
+```dart
+class Box<T> {
+  final T? object;
+  Box.empty();
+  Box.full(this.object);
+}
+```
+
+Note the `?` on the declaration of `object`. Now the field has an explicitly
+nullable type, so it is fine to leave it uninitialized.
+
+When you make a type parameter type nullable like `T?` here, you may need to
+cast the nullability away. The correct way to do that is using an explicit `as
+T` cast, *not* the `!` operator:
+
+```dart
+class Box<T> {
+  final T? object;
+  Box.empty();
+  Box.full(this.object);
+
+  T unbox() => object as T;
+}
+```
+
+The `!` operator *always* throws if the value is `null`. But if the type
+parameter has been instantiated with a nullable type, then `null` is a perfectly
+valid value for `T`:
+
+```dart
+main() {
+  var box = Box<int?>.full(null);
+  print(box.unbox());
+}
+```
+
+This program should run without error. Using `as T` accomplishes that. Using
+`!` would throw an exception.
+
+Other generic types have some bound that restricts the kinds of type arguments
+that can be applied:
+
+```dart
+Rect<T extends num> {
```

Comment:
`class Rect`?

Reply:
You mean we aren't making `class` optional? :) Done.